### DISCUSSION TITLE
Storage engine interface

### DISCUSSION THREAD

**Discussion by leiysky** [korandoru/zeronos#3] [General]
Seems `dryad` is a distributed key-value storage. I'd like to know about the design of storage engine.

**Comment by tisonkun** [comment]
@leiysky currently I build the system upon [Apache Ratis](https://github.com/apache/ratis). Ratis provides storage of Raft logs and uses snapshot storage provided by application. You may read the source code of Ratis to learn how it writes Raft logs, and we use the `SimpleStateMachineStorage` to store snapshot (versioned by term) in one file.

Perhaps I can add a config of base data dir so that you can look at the directories and files layout (#4).

**Reply by tisonkun** [reply]
So far, we store data by raft groups.

```
└── 02511d47-d67c-49a3-9011-abb3109a44c1 # group-id
    ├── current # raft logs
    │   ├── log_inprogress_0
    │   ├── raft-meta
    │   └── raft-meta.conf
    ├── in_use.lock
    └── sm # state machine
```

@leiysky you may read more from `RaftStorageDirectory`'s comment. Ratis doesn't provide rich document online but almost all in its comments.
